ntp-udp/src/lib.rs

```diff
@@ -0,0 +1,128 @@
+use std::{
+    io,
+    io::{ErrorKind, IoSliceMut},
+    os::unix::prelude::AsRawFd,
+};
+
+use ntp_proto::NtpTimestamp;
+use tokio::io::unix::AsyncFd;
+
+pub struct UdpSocket {
+    io: AsyncFd<std::net::UdpSocket>,
+}
+
+impl UdpSocket {
+    pub fn from_std(socket: std::net::UdpSocket) -> io::Result<UdpSocket> {
+        socket.set_nonblocking(true)?;
+        init_socket(&socket)?;
+        Ok(UdpSocket {
+            io: AsyncFd::new(socket)?,
+        })
+    }
+
+    pub fn from_tokio(socket: tokio::net::UdpSocket) -> io::Result<UdpSocket> {
+        // tokio sockets are already non-blocking
+        let socket = socket.into_std()?;
+        init_socket(&socket)?;
+        Ok(UdpSocket {
+            io: AsyncFd::new(socket)?,
+        })
+    }
+
+    pub async fn send(&self, buf: &[u8]) -> io::Result<usize> {
+        loop {
+            let mut guard = self.io.writable().await?;
+            match guard.try_io(|inner| inner.get_ref().send(buf)) {
+                Ok(result) => return result,
+                Err(_would_block) => continue,
+            }
+        }
+    }
+
+    pub async fn recv(
+        &self,
+        buf: &mut [u8],
+        recv_ts: &mut Option<NtpTimestamp>,
+    ) -> io::Result<usize> {
+        loop {
+            let mut guard = self.io.readable().await?;
+            match guard.try_io(|inner| recv(inner.get_ref(), buf, recv_ts)) {
+                Ok(result) => return result,
+                Err(_would_block) => continue,
+            }
+        }
+    }
+}
+
+fn init_socket(socket: &std::net::UdpSocket) -> io::Result<()> {
+    let fd = socket.as_raw_fd();
+    let enable_ts: libc::c_int = 1;
+    unsafe {
+        libc::setsockopt(
+            fd,
+            libc::SOL_SOCKET,
+            libc::SO_TIMESTAMPNS_NEW,
+            &enable_ts as *const _ as *const libc::c_void,
+            std::mem::size_of::<libc::c_int>() as libc::socklen_t,
+        )
+    };
+    Ok(())
+}
+
+fn prepare_header(mhdr: &mut libc::msghdr, buf: &mut IoSliceMut, control_buf: &mut [u8]) {
+    mhdr.msg_control = control_buf as *mut _ as *mut libc::c_void;
+    mhdr.msg_controllen = control_buf.len();
+    mhdr.msg_iov = buf as *mut _ as *mut _;
+    mhdr.msg_iovlen = 1;
+    mhdr.msg_flags = 0;
+}
+
+fn recv(
+    socket: &std::net::UdpSocket,
+    buf: &mut [u8],
+    recv_ts: &mut Option<NtpTimestamp>,
```

Comment:
Using a `&mut Option` as an outparam seems pretty unidiomatic. I think it would definitely be better to return a tuple (why do we think callers will be interested in the length of the read?).

Reply:
Definitely! So the final version already changed that! The length of the read would mostly be important to ensure that the buffer was filled with the expected number of bytes and not fewer than expected.